atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/Transaction.java

```diff
@@ -41,25 +41,95 @@
+     * array, as well as with byte arrays that are equal in terms of {@link java.util.Arrays#equals(byte[], byte[])}.
+     *
+     * If there are rows with no cells matching the provided {@link ColumnSelection}, they will not be present in the
+     * {@link Map#keySet()} of the output map at all.
+     *
+     * @param tableRef table to load rows from
+     * @param rows rows to be loaded
+     * @param columnSelection columns to load from the given rows
+     * @return a mapping of rows to the columns matching the provided column selection
+     */
     @Idempotent
     SortedMap<byte[], RowResult<byte[]>> getRows(
             TableReference tableRef,
             Iterable<byte[]> rows,
             ColumnSelection columnSelection);
 
+    /**
+     * Returns a mapping of rows to {@link BatchingVisitable}s (which may be thought of as iterators over columns)
+     * within {@code tableRef} for the specified {@code rows}, where the columns fall within the provided
+     * {@link BatchColumnRangeSelection}. The single provided {@link BatchColumnRangeSelection} applies to all of the
+     * rows.
+     *
+     * The returned {@link BatchingVisitable}s are guaranteed to return cells matching the predicate, sorted on
+     * (ascending) byte ordering.
+     *
+     * It is guaranteed that the {@link Map#keySet()} of the returned map has the same elements as {@code rows} by
+     * identity, even if there are rows where no columns match the predicate. Random access for the returned map
+     * should ONLY be performed through the original byte arrays that were passed in.
+     *
+     * @param tableRef table to load values from
+     * @param rows unique rows to apply the column range selection to
+     * @param columnRangeSelection range of columns and batch size to load for each of the rows provided
+     * @return a mapping of rows to cells matching the predicate in the row, following the ordering outlined above
+     * @throws IllegalArgumentException if {@code rows} contains duplicates
+     */
     @Idempotent
     Map<byte[], BatchingVisitable<Map.Entry<Cell, byte[]>>> getRowsColumnRange(
             TableReference tableRef,
             Iterable<byte[]> rows,
             BatchColumnRangeSelection columnRangeSelection);
 
+    /**
+     * Returns a single iterator over the cell-value pairs in {@code tableRef} for the specified {@code rows}, where the
+     * columns fall within the provided {@link ColumnRangeSelection}. The single provided {@link ColumnRangeSelection}
+     * applies to all of the rows.
+     *
+     * If the provided {@link Iterable} of {@code rows} has a stable ordering, the returned iterator is guaranteed
+     * to return cell-value pairs in a lexicographic ordering over rows and columns, where rows are sorted according to
+     * the stable ordering of {@code rows}, and columns are sorted on byte ordering. If {@code rows} does not have a
+     * stable ordering, behaviour is undefined.
```

Comment:
Can we tighten this up as well? You're right, of course, but all hell breaks loose if this happens. So when you see something like this, I think it's on us to actually clean the contract. In this case, internally we should do ImmutableList.copyOf.

Reply:
(assuming that this is not lazy, in which case the contract might need tidying up).

Reply:
Yeah, added an internal ImmutableList.copyOf, and also a funky test that actually gives unstable iteration order.